assets/js/modules/tagmanager/components/setup/SetupMain.js

```diff
@@ -50,29 +50,23 @@ export default function SetupMain( { finishSetup } ) {
 	const isDoingGetAccounts = useSelect( ( select ) => select( STORE_NAME ).isDoingGetAccounts() );
 	const isDoingSubmitChanges = useSelect( ( select ) => select( STORE_NAME ).isDoingSubmitChanges() );
 	const hasResolvedAccounts = useSelect( ( select ) => select( STORE_NAME ).hasFinishedResolution( 'getAccounts' ) );
+	const inProgress = useSelect( ( select ) => select( CORE_FORMS ).getValue( FORM_SETUP, 'inProgress' ) );
 	const isCreateAccount = ACCOUNT_CREATE === accountID;
 
 	// Set the accountID and containerID if there is an existing tag.
 	useExistingTagEffect();
 
-	// When `finishSetup` is called, flag that we are navigating to keep the progress bar going.
-	const [ isNavigating, setIsNavigating ] = useState( false );
-	const finishSetupAndNavigate = ( ...args ) => {
-		finishSetup( ...args );
-		setIsNavigating( true );
-	};
```

Comment:
Not strictly necessary the way we have our code behave now, but I think it would be good to keep this functionality as it was before, just using `dispatch` to set `inProgress` in `core/forms` store instead. Then pass the full `finishSetupAndNavigate` to the `SetupForm` component below.

This way the `finishSetup` prop adds the progress bar on its own as expected when called, and `SetupForm` can still do its adjustments as needed. Due to this component and the progress bar now relying on a single shared value, despite potentially setting it in multiple places it wouldn't have any drawbacks regarding appearance.

Reply:
While this is more/less a harmless change, I'm having a hard time seeing why we would keep this over the simpler implementation. `finishSetup` in the `SetupForm` is only ever called within its `submitForm` function which handles the now `submitInProgress` form state as well as the form submission itself in all cases. Why would we modify this function in a way that wouldn't make a difference? If this function were called somewhere else, then I think this would make a lot more sense and would even be necessary, but I'm hesitant to make the change as-is. Maybe I'm missing something though – can you elaborate on why this would be good to keep?

Reply:
I think keeping this in here too means that the `finishSetup` function is more self-contained. The other component that receives it as prop and calls it shouldn't _need_ to deal with `submitInProgress`, it's only a UX optimization that makes sense in the case of Tag Manager.

Yes, for our current implementation it is not important to have but I think it's more bulletproof to ensure that, no matter what the other component does, `finishSetup` would behave like it should.

Reply:
> I think keeping this in here too means that the `finishSetup` function is more self-contained. The other component that receives it as prop and calls it shouldn't _need_ to deal with `submitInProgress` [...]

In that sense, if we were to bring it back, I think it makes more sense to keep the local `isNavigating` state which is what `finishSetup` would set when called and add that back as an additional condition for showing the progress bar. It doesn't make sense for `finishSetup` to set `submitInProgress` because by the time it's called, the submit is basically done. Maybe we can update the `SetupForm` to always set `submitInProgress` to `false` when it's done, and we would rely on the state set when calling `finishSetup` to keep the progress bar going like it did before. I don't think there should be any flicker in that case as `submitInProgress` would still be `true` when `finishSetup` is called. What do you think?

> it's only a UX optimization that makes sense in the case of Tag Manager

I think the pattern here applies to other modules as well, but we can look at them as-needed.

Reply:
@felixarntz this has been updated as we talked about now. Just waiting for the build to finish – seeing some failures due to the Composer image failing to pull so it may need a few of them restarted. PHP and JS tests are all passing for me locally though.